src/model_signing/_serialization/incremental.py

```diff
@@ -0,0 +1,247 @@
+            ignore_paths: The paths to ignore during serialization. If a
+              provided path is a directory, all children of the directory are
+              ignored.
+            files_to_hash: Optional list of files that may have changed and
+              should be re-hashed. If None, all files in the model directory
+              are scanned, and only NEW files (not in existing manifest) are
+              hashed. Existing files have their digests reused.
+
+              To detect changed files, use git diff or similar:
+                  changed_files = subprocess.check_output(
+                      ['git', 'diff', '--name-only', 'HEAD']
+                  ).decode().splitlines()
+                  files_to_hash = [model_path / f for f in changed_files]
+
+        Returns:
+            The model's serialized manifest with a mix of reused and
+            newly-computed digests.
+
+        Raises:
+            ValueError: The model contains a symbolic link, but the serializer
+              was not initialized with `allow_symlinks=True`.
+        """
+        # Build a set of files to rehash (files that potentially changed)
+        rehash_paths = set()
+        if files_to_hash is not None:
+            # User provided explicit list of changed files
+            for path in files_to_hash:
+                if path.is_file():
+                    rehash_paths.add(path.relative_to(model_path))
+
+        # Scan directory to find all current files in the model
+        all_current_files = []
+        for path in itertools.chain((model_path,), model_path.glob("**/*")):
+            if serialization.should_ignore(path, ignore_paths):
+                continue
+            serialization.check_file_or_directory(
+                path, allow_symlinks=self._allow_symlinks
+            )
+            if path.is_file():
+                all_current_files.append(path)
+
+        # Build the new manifest
+        files_to_rehash = []
+        manifest_items = []
+
+        for path in all_current_files:
+            relative_path = path.relative_to(model_path)
+            posix_path = pathlib.PurePosixPath(relative_path)
+
+            # Determine if this file needs re-hashing
+            needs_rehash = False
+
+            if posix_path not in self._existing_items:
+                # New file not in old manifest - must hash it
+                needs_rehash = True
+            elif rehash_paths and relative_path in rehash_paths:
+                # File was explicitly marked as changed - must re-hash it
+                needs_rehash = True
+            elif not rehash_paths:
+                # No explicit files_to_hash provided, so we're in "scan mode"
+                # Reuse digest for existing files (assume unchanged)
+                needs_rehash = False
+            else:
+                # File exists in old manifest and wasn't marked as changed
+                # Reuse old digest
+                needs_rehash = False
```

Comment:
This block should be extracted to own utility function (since it will likely be reused when adding support for incremental hashing of shards). It also improves readability, as this function right now is quite large.